core/capabilities/webapi/outgoing_connector_handler.go

```
@@ -92,58 +85,94 @@ func (c *OutgoingConnectorHandler) HandleSingleNodeRequest(ctx context.Context,
 	// Create a subcontext with the timeout plus some margin for the gateway to process the request
 	timeoutDuration := time.Duration(req.TimeoutMs) * time.Millisecond
 	margin := 100 * time.Millisecond
-	ctx, cancel := context.WithTimeout(ctx, timeoutDuration+margin)
+	ctxWithTimeout, cancel := context.WithTimeout(ctx, timeoutDuration+margin)
 	defer cancel()
 
-	payload, err := json.Marshal(req)
-	if err != nil {
-		return nil, fmt.Errorf("failed to marshal fetch request: %w", err)
-	}
+	sendRequest := c.createSendRequest(messageID, req)
 
-	ch, err := c.responses.new(messageID)
-	if err != nil {
-		return nil, fmt.Errorf("duplicate message received for ID: %s", messageID)
+	return c.handleSingleNodeRequest(ctxWithTimeout, sendRequest, req.MaxRetries)
+}
+
+// handleSingleNodeRequest accepts a blocking function that returns an api message and error and either calls it directly or with
+// a retry strategy.
+func (c *OutgoingConnectorHandler) handleSingleNodeRequest(
+	ctx context.Context,
+	sendRequest func(context.Context) (*api.Message, error),
+	maxRetries uint32,
+) (*api.Message, error) {
+	if maxRetries > 0 {
+		strategy := &retry.Strategy[*api.Message]{
+			MaxRetries: uint(min(maxRetries, defaultMaxRetries)),
+		}
+		return strategy.Do(ctx, c.lggr, sendRequest)
 	}
-	defer c.responses.cleanup(messageID)
 
-	lggr.Debugw("sending request to gateway")
+	return sendRequest(ctx)
+}
 
-	body := &api.MessageBody{
-		MessageId: messageID,
-		DonId:     c.gc.DonID(),
-		Method:    c.method,
-		Payload:   payload,
-	}
+// createSendRequest closes over a message ID and request to create a function that blocks on call
+// until a message and error are returned.
+func (c *OutgoingConnectorHandler) createSendRequest(messageID string, req capabilities.Request) func(context.Context) (*api.Message, error) {
```

🤔 IIRC the original intent behind this is for users to be able to specify when they want to retry requests, but in what circumstances?

Looking at the below, it looks like we can return an error in the following cases:
- rate limit -> we probably shouldn't retry then
- connection failure to gateway -> we should retry (with the next gateway)
- internal error -> not sure if we should retry? I think this mainly indicates a rate limit on the upstream gateway, in which case we shouldn't retry
- ctx.Done -> we should retry, but since the context is global it won't actually do anything

This leaves two questions in my mind:
- What about errors from the actual API we are calling? Should we retry if this returns a non-2xx response? At the moment we aren't
- Should we be better distinguishing between retryable and non-retryable errors?

Wdyt @MStreet3 ?

---

good points, thinking on it.  The way I've structured this has conflated some issues.